File: packages/contracts/src/risk-checks.ts
import { z } from "zod";

import { IdentifierSchema, IsoDateTimeSchema, UuidSchema } from "./primitives";

export const RISK_CHECK_STATES = ["pass", "warn", "fail", "na"] as const;

export const RiskCheckStateSchema = z.enum(RISK_CHECK_STATES);
export type RiskCheckState = z.infer<typeof RiskCheckStateSchema>;

const RiskCheckDetailSchema = z.string().trim().min(1).max(512);

/** A deterministic risk result. It is read-only evidence, never an override. */
export const RiskCheckSchema = z
  .object({
    key: IdentifierSchema,
    labelTh: IdentifierSchema,
    labelEn: IdentifierSchema,
    state: RiskCheckStateSchema,
    actual: RiskCheckDetailSchema,
    limit: RiskCheckDetailSchema.nullish(),
    hard: z.boolean(),
  })
  .strict();

export type RiskCheck = z.infer<typeof RiskCheckSchema>;

/** Persistence-boundary shape used when a risk check is read from the control plane. */
export const PersistedRiskCheckSchema = RiskCheckSchema.omit({
  limit: true,
})
  .extend({
    id: UuidSchema,
    ownerId: UuidSchema,
    tradeProposalId: UuidSchema,
    proposalVersion: z.number().int().positive(),
    limitValue: RiskCheckDetailSchema.nullable(),
    explanation: RiskCheckDetailSchema.nullable(),
    ordinal: z.number().int().nonnegative(),
    createdAt: IsoDateTimeSchema,
  })
  .strict();

export type PersistedRiskCheck = z.infer<typeof PersistedRiskCheckSchema>;

export function riskCheckFromPersisted(record: PersistedRiskCheck): RiskCheck {
  return RiskCheckSchema.parse({
    key: record.key,
    labelTh: record.labelTh,
    labelEn: record.labelEn,
    state: record.state,
    actual: record.actual,
    limit: record.limitValue,
    hard: record.hard,
  });
}

File: apps/worker/src/aurum_worker/models/risk_checks.py
"""Read-only deterministic risk-check contracts."""

from __future__ import annotations

from enum import StrEnum
from typing import Annotated
from uuid import UUID

from pydantic import AwareDatetime, StringConstraints

from aurum_worker.models.base import Identifier, NonNegativeInt, PositiveInt, WireModel

RiskCheckDetail = Annotated[
    str,
    StringConstraints(
        strip_whitespace=True,
        strict=True,
        min_length=1,
        max_length=512,
    ),
]


class RiskCheckState(StrEnum):
    PASS = "pass"
    WARN = "warn"
    FAIL = "fail"
    NA = "na"


class RiskCheck(WireModel):
    """Deterministic evidence only; this model grants no override."""

    key: Identifier
    label_th: Identifier
    label_en: Identifier
    state: RiskCheckState
    actual: RiskCheckDetail
    limit: RiskCheckDetail | None = None
    hard: bool


class PersistedRiskCheck(WireModel):
    id: UUID
    owner_id: UUID
    trade_proposal_id: UUID
    proposal_version: PositiveInt
    key: Identifier
    label_th: Identifier
    label_en: Identifier
    state: RiskCheckState
    actual: RiskCheckDetail
    limit_value: RiskCheckDetail | None
    explanation: RiskCheckDetail | None
    hard: bool
    ordinal: NonNegativeInt
    created_at: AwareDatetime

    def to_risk_check(self) -> RiskCheck:
        return RiskCheck.model_validate(
            {
                "key": self.key,
                "labelTh": self.label_th,
                "labelEn": self.label_en,
                "state": self.state,
                "actual": self.actual,
                "limit": self.limit_value,
                "hard": self.hard,
            }
        )
